spring-integration-core/src/main/java/org/springframework/integration/codec/CompositeCodec.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.codec;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Map;
+
+import org.springframework.integration.util.ClassUtils;
+import org.springframework.util.Assert;
+
+
+/**
+ * A Codec that can delegate to one out of many Codecs, each mapped to a class.
+ * @author David Turanski
+ * @since 4.2
+ */
+public class CompositeCodec implements Codec {
+
+	private final Codec defaultCodec;
+
+	private final Map<Class<?>, Codec> delegates;
+
+	public CompositeCodec(Map<Class<?>, Codec> delegates, Codec defaultCodec) {
+		Assert.notNull(defaultCodec, "'defaultCodec' cannot be null");
+		this.defaultCodec = defaultCodec;
+		this.delegates = delegates;
```

Comment:
`new HashMap(delegates)` - don't keep a reference to the caller's map - he might change it.